dev/integration_tests/external_ui/test_driver/main_test.dart

```diff
@@ -0,0 +1,55 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'package:flutter_driver/flutter_driver.dart';
+import 'package:test/test.dart';
+
+final RegExp statsRegExp = new RegExp('Produced: (.*)fps\nConsumed: (.*)fps\nWidget builds: (.*)');
+
+Future<Null> main() async {
+
+  // TODO(mravn): the following pause appears necessary on iOS to avoid
+  // inflating elements too early (when there is no size).
+  await new Future<Null>.delayed(const Duration(seconds: 1));
+
+  group('texture suite', () {
+    FlutterDriver driver;
+
+    setUpAll(() async {
+      driver = await FlutterDriver.connect();
+    });
+
+    test('texture rendering', () async {
+      final SerializableFinder fab = find.byValueKey('fab');
+      final SerializableFinder summary = find.byValueKey('summary');
+
+      await driver.tap(fab);
+      await new Future<Null>.delayed(const Duration(seconds: 3));
+      await driver.tap(fab);
+
+      final String statsSlow = await driver.getText(summary);
+      final Match matchSlow = statsRegExp.matchAsPrefix(statsSlow);
+      expect(matchSlow, isNotNull);
+      expect(double.parse(matchSlow.group(1)), closeTo(10.0, 2.0));
+      expect(double.parse(matchSlow.group(2)), closeTo(10.0, 2.0));
+      expect(int.parse(matchSlow.group(3)), 1);
+
+      await driver.tap(fab);
+      await new Future<Null>.delayed(const Duration(seconds: 3));
+      await driver.tap(fab);
+
+      final String statsFast = await driver.getText(summary);
+      final Match matchFast = statsRegExp.matchAsPrefix(statsFast);
+      expect(matchFast, isNotNull);
+      expect(double.parse(matchFast.group(1)), closeTo(100.0, 2.0));
+      expect(double.parse(matchFast.group(2)), inInclusiveRange(52.0, 61.0));
```

Comment:
Are these expectations measuring performance or correctness?

If correctness, then how resilient are they to device type (e.g. low-end Android devices are capped at 30fps, while some very new ones are going 120fps).

If performance, then instead of failing the test we should submit the numbers as benchmark scores to be tracked on the dashboard ([example](https://github.com/flutter/flutter/blob/2a11e896eef48d503841b4c8c4a83638e2d8e5f6/dev/devicelab/lib/tasks/perf_tests.dart#L112)).

Reply:
Thanks for the review. Both comments addressed. The expectations are about correctness: we want texture consumption to proceed at rate min(flutterFrameRate, textureProductionFrameRate). And we want to ensure that the Widget tree is not rebuilt, if nothing happens except texture rendering. I've added code to calibrate Flutter's frame rate on app launch, then I produce texture frames at 0.5x and 2x that rate and check that they are consumed at roughly 0.5x and 1x Flutter's frame rate.